lib/src/secure_config.rs

```diff
@@ -0,0 +1,571 @@
+    // If path_to_bytes failed, jj should not stop working.
+    // Instead, we simply treat all non-utf8 paths as empty.
+    fn encoded_path(&self) -> &[u8] {
+        path_to_bytes(&self.repo_dir).ok().unwrap_or_default()
+    }
+
+    fn generate_config_id(&self, rng: &RefCell<ChaCha20Rng>) -> String {
+        encode_hex(
+            &(0..CONFIG_ID_BYTES)
+                .map(|_| rng.borrow_mut().random::<u8>())
+                .collect::<Vec<_>>(),
+        )
+    }
+
+    fn generate_config(
+        &self,
+        config_dir: &Path,
+        config_id: &str,
+        content: &[u8],
+        metadata: &ConfigMetadata,
+    ) -> Result<PathBuf, SecureConfigError> {
+        let config_dir = config_dir.join(config_id);
+        let config_path = config_dir.join(CONFIG_FILE);
+        fs::create_dir_all(&config_dir).context(&config_dir)?;
+        self.update_metadata(&config_dir, metadata)?;
+        if !content.is_empty() {
+            fs::write(&config_path, content).context(&config_path)?;
+        }
+
+        // Write the config ID atomically. A half-formed config ID would be very bad.
+        atomic_write(
+            &self.repo_dir.join(self.config_id_name),
+            config_id.as_bytes(),
+        )?;
+        Ok(config_path)
+    }
+
+    fn generate_initial_config(
+        &self,
+        config_dir: &Path,
+        config_id: &str,
+    ) -> Result<(PathBuf, ConfigMetadata), SecureConfigError> {
+        let metadata = ConfigMetadata {
+            path: self.encoded_path().to_vec(),
+        };
+        let path = self.generate_config(config_dir, config_id, &[], &metadata)?;
+        Ok((path, metadata))
+    }
+
+    fn update_metadata(
+        &self,
+        config_dir: &Path,
+        metadata: &ConfigMetadata,
+    ) -> Result<(), SecureConfigError> {
+        let metadata_path = config_dir.join(METADATA_FILE);
+        atomic_write(&metadata_path, &metadata.encode_to_vec())?;
+        Ok(())
+    }
+
+    /// Validates that the metadata path matches the repo path.
+    /// If there's a mismatch, takes appropriate action.
+    /// Returns the updated config dir and metadata.
+    fn handle_metadata_path(
+        &self,
+        rng: &RefCell<ChaCha20Rng>,
+        root_config_dir: &Path,
+        config_dir: PathBuf,
+        mut metadata: ConfigMetadata,
+    ) -> Result<(PathBuf, ConfigMetadata), SecureConfigError> {
+        let (encoded, want) = if let Ok(encoded) = path_to_bytes(&self.repo_dir) {
+            (encoded, Some(self.repo_dir.as_path()))
+        } else {
+            Default::default()
+        };
+        let got = if metadata.path.is_empty() {
+            None
+        } else {
+            Some(path_from_bytes(&metadata.path)?)
+        };
+
+        if want == got {
+            return Ok((config_dir, metadata));
+        }
+        let got = match got {
+            Some(x) if x.is_dir() => x,
+            _ => {
+                // The old repo does not exist. Assume the user moved it.
```

Comment:
I think we should print a message when copy/rename gets detected and config location is updated accordingly.

Maybe we can return stats enum, struct, or list of enums and let caller print it?

Reply:
I've done this for copies. I won't do this for renames, however, for two reasons:
* Nothing changes from a user's perspective after a rename. The config file remains at the same location, the only thing that gets updated is metadata.binpb, which the user doesn't see.
* Renames can have a lot more false positives than copies. For example
  * In a dual-boot system, every time I switch between linux and windows
  * Every time I remount an external hard drive at a different location
  * Every time I run a command inside a chroot.